hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/streaming/BlockBufferService.java

```diff
@@ -0,0 +1,604 @@
+// SPDX-License-Identifier: Apache-2.0
+package com.hedera.node.app.blocks.impl.streaming;
+
+import static com.hedera.node.app.blocks.impl.streaming.BlockBufferService.BlockStreamQueueItemType.BLOCK_ITEM;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.block.stream.BlockItem;
+import com.hedera.node.app.metrics.BlockStreamMetrics;
+import com.hedera.node.config.ConfigProvider;
+import com.hedera.node.config.data.BlockStreamConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Manages the state and lifecycle of blocks being streamed to block nodes.
+ * This class is responsible for:
+ * <ul>
+ *     <li>Maintaining the block states in a buffer</li>
+ *     <li>Handling backpressure when the buffer is saturated</li>
+ *     <li>Pruning the buffer based on TTL and saturation</li>
+ * </ul>
+ */
+public class BlockBufferService {
+    private static final Logger logger = LogManager.getLogger(BlockBufferService.class);
+
+    /**
+     * Buffer that stores recent blocks. This buffer is unbounded, however when opening a new block the buffer will be
+     * pruned. Generally speaking, the buffer should contain only blocks that are recent (that is within the configured
+     * {@link BlockStreamConfig#blockBufferTtl() TTL}) and have yet to be acknowledged. There may be cases where older
+     * blocks still exist in the buffer if they are unacknowledged, but once they are acknowledged they will be pruned
+     * the next time {@link #openBlock(long)} is invoked.
+     */
+    private final Queue<BlockState> blockBuffer = new ConcurrentLinkedQueue<>();
```

Comment:
I don't think it's immediately obvious why this structure is necessary, since `ConcurrentHashMap` also offers thread-safe iteration through its entries.